File: internal/doctor/doctor.go
package doctor

import (
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"strings"

	"github.com/mauricedesaxe/lazar-harness/internal/resolve"
)

type Status string

const (
	OK   Status = "ok"
	WARN Status = "warn"
	INFO Status = "info"
)

type Check struct {
	Name   string `json:"name"`
	Status Status `json:"status"`
	Detail string `json:"detail"`
}

type Report struct {
	Root   string  `json:"root"`
	Checks []Check `json:"checks"`
}

func (r Report) String() string {
	var b strings.Builder
	fmt.Fprintf(&b, "harness-check doctor: %s\n", r.Root)
	for _, c := range r.Checks {
		fmt.Fprintf(&b, "%-4s %s: %s\n", strings.ToUpper(string(c.Status)), c.Name, c.Detail)
	}
	return b.String()
}

func Run(root string, opts resolve.Options) Report {
	cfgs := resolve.Detect(root)
	var checks []Check
	checks = append(checks, ruffChecks(root, cfgs, opts)...)
	checks = append(checks, oxlintChecks(root, cfgs, opts)...)
	checks = append(checks, toolCheck("ruff", orDefault(opts.RuffBin, "ruff")), toolCheck("oxlint", orDefault(opts.OxlintBin, "oxlint")))
	return Report{Root: root, Checks: checks}
}

func orDefault(bin, fallback string) string {
	if bin == "" {
		return fallback
	}
	return bin
}

func ruffChecks(root string, cfgs resolve.RepoConfigs, opts resolve.Options) []Check {
	var out []Check
	if cfgs.RuffConfig == "" {
		return append(out, Check{
			Name: "ruff-config", Status: WARN,
			Detail: "no repo ruff config found; the harness baseline applies at edit time. Extend it to own this repo's policy.",
		})
	}
	if extendsBaseline(cfgs.RuffConfig) {
		out = append(out, Check{Name: "ruff-config", Status: OK, Detail: "extends the harness baseline"})
	} else {
		out = append(out, Check{
			Name: "ruff-config", Status: WARN,
			Detail: "does not extend the harness baseline; add extend = \"<baseline>/ruff.toml\" to inherit updates",
		})
	}
	missing := missingFamilies(cfgs.RuffConfig, opts)
	if len(missing) > 0 {
		out = append(out, Check{
			Name: "ruff-families", Status: WARN,
			Detail: "baseline families missing from repo config: " + strings.Join(missing, ", "),
		})
	} else {
		out = append(out, Check{Name: "ruff-families", Status: OK, Detail: "covers all baseline families"})
	}
	return out
}

func oxlintChecks(root string, cfgs resolve.RepoConfigs, opts resolve.Options) []Check {
	var out []Check
	if cfgs.OxlintConfig == "" {
		return append(out, Check{
			Name: "oxlint-config", Status: WARN,
			Detail: "no .oxlintrc.json found; the harness baseline applies at edit time. Extend it to own this repo's policy.",
		})
	}
	if extendsBaseline(cfgs.OxlintConfig) {
		out = append(out, Check{Name: "oxlint-config", Status: OK, Detail: "extends the harness baseline"})
	} else {
		out = append(out, Check{
			Name: "oxlint-config", Status: WARN,
			Detail: "does not extend the harness baseline; add it under extends to inherit updates",
		})
	}
	return out
}

func toolCheck(name string, bin string) Check {
	path, err := exec.LookPath(bin)
	if err != nil {
		return Check{Name: name, Status: INFO, Detail: bin + " not on PATH; edit-time " + name + " checks are skipped"}
	}
	out, err := exec.Command(bin, "--version").Output()
	detail := path
	if err == nil {
		detail = strings.TrimSpace(string(out))
	}
	return Check{Name: name, Status: OK, Detail: detail}
}

var extendHint = regexp.MustCompile(`lazar-harness/linters|lint-baselines`)

func extendsBaseline(configPath string) bool {
	b, err := os.ReadFile(configPath)
	return err == nil && extendHint.Match(b)
}

var selectRe = regexp.MustCompile(`select\s*=\s*\[([^\]]*)\]`)

// parseSelects is a best-effort reader for the ruff `select` list. It strips
// TOML line comments first, because real configs annotate every family, then
// reads the plain quoted-name subset.
func parseSelects(configPath string) []string {
	b, err := os.ReadFile(configPath)
	if err != nil {
		return nil
	}
	var noComments strings.Builder
	for _, line := range strings.Split(string(b), "\n") {
		if idx := strings.Index(line, "#"); idx >= 0 {
			line = line[:idx]
		}
		noComments.WriteString(line)
		noComments.WriteString(" ")
	}
	m := selectRe.FindStringSubmatch(noComments.String())
	if m == nil {
		return nil
	}
	var out []string
	for _, part := range strings.Split(m[1], ",") {
		name := strings.Trim(strings.TrimSpace(part), "\"'")
		if name != "" {
			out = append(out, name)
		}
	}
	return out
}

func missingFamilies(repoConfig string, opts resolve.Options) []string {
	repo := parseSelects(repoConfig)
	if len(repo) == 0 {
		return nil
	}
	base := parseSelects(resolve.BaselinePath(opts, "ruff.toml"))
	have := make(map[string]bool, len(repo))
	for _, f := range repo {
		have[f] = true
	}
	var missing []string
	for _, f := range base {
		if !have[f] {
			missing = append(missing, f)
		}
	}
	return missing
}

File: src/doctor.rs
use serde::Serialize;

use crate::resolve;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum Status {
    Ok,
    Warn,
    Info,
}

#[derive(Debug, Serialize)]
pub struct Check {
    pub name: String,
    pub status: Status,
    pub detail: String,
}

#[derive(Debug, Serialize)]
pub struct Report {
    pub root: String,
    pub checks: Vec<Check>,
}

impl std::fmt::Display for Status {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Status::Ok => write!(f, "ok"),
            Status::Warn => write!(f, "warn"),
            Status::Info => write!(f, "info"),
        }
    }
}

impl std::fmt::Display for Report {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "harness-check doctor: {}", self.root)?;
        for c in &self.checks {
            writeln!(
                f,
                "{:<4} {}: {}",
                format!("{}", c.status).to_uppercase(),
                c.name,
                c.detail
            )?;
        }
        Ok(())
    }
}

pub fn run(root: &str, opts: &resolve::Options) -> Report {
    let cfgs = resolve::detect(root);
    let mut checks = vec![];
    checks.push(baseline_installed(opts));
    checks.extend(ruff_checks(&cfgs, opts));
    checks.extend(oxlint_checks(&cfgs, opts));
    checks.push(tool_check("ruff", opts.ruff_bin().to_string()));
    checks.push(tool_check("oxlint", opts.oxlint_bin().to_string()));
    Report {
        root: root.into(),
        checks,
    }
}

fn baseline_installed(opts: &resolve::Options) -> Check {
    match resolve::validate_baseline(opts) {
        Ok(()) => Check {
            name: "harness-baseline".into(),
            status: Status::Ok,
            detail: opts.baseline_dir(),
        },
        Err(e) => Check {
            name: "harness-baseline".into(),
            status: Status::Warn,
            detail: e,
        },
    }
}

fn ruff_checks(cfgs: &resolve::RepoConfigs, opts: &resolve::Options) -> Vec<Check> {
    let Some(config) = &cfgs.ruff_config else {
        return vec![Check {
            name: "ruff-config".into(),
            status: Status::Warn,
            detail: "no repo ruff config found; the harness baseline applies at edit time. Extend it to own this repo's policy.".into(),
        }];
    };
    let mut out = vec![];
    if extends_baseline(config) {
        out.push(Check {
            name: "ruff-config".into(),
            status: Status::Ok,
            detail: "extends the harness baseline".into(),
        });
    } else {
        out.push(Check {
            name: "ruff-config".into(),
            status: Status::Warn,
            detail: "does not extend the harness baseline; add extend = \"<baseline>/ruff.toml\" to inherit updates".into(),
        });
    }
    let missing = missing_families(config, opts);
    if missing.is_empty() {
        out.push(Check {
            name: "ruff-families".into(),
            status: Status::Ok,
            detail: "covers all baseline families".into(),
        });
    } else {
        out.push(Check {
            name: "ruff-families".into(),
            status: Status::Warn,
            detail: format!(
                "baseline families missing from repo config: {}",
                missing.join(", ")
            ),
        });
    }
    out
}

fn oxlint_checks(cfgs: &resolve::RepoConfigs, _opts: &resolve::Options) -> Vec<Check> {
    let Some(config) = &cfgs.oxlint_config else {
        return vec![Check {
            name: "oxlint-config".into(),
            status: Status::Warn,
            detail: "no .oxlintrc.json found; the harness baseline applies at edit time. Extend it to own this repo's policy.".into(),
        }];
    };
    if extends_baseline(config) {
        vec![Check {
            name: "oxlint-config".into(),
            status: Status::Ok,
            detail: "extends the harness baseline".into(),
        }]
    } else {
        vec![Check {
            name: "oxlint-config".into(),
            status: Status::Warn,
            detail: "does not extend the harness baseline; add it under extends to inherit updates"
                .into(),
        }]
    }
}

fn tool_check(name: &str, bin: String) -> Check {
    if !crate::resolve::which_bin(&bin) {
        return Check {
            name: name.into(),
            status: Status::Info,
            detail: format!("{bin} not on PATH; edit-time {name} checks are skipped"),
        };
    }
    let detail = std::process::Command::new(&bin)
        .arg("--version")
        .output()
        .ok()
        .map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string())
        .filter(|s| !s.is_empty())
        .unwrap_or_else(|| bin.clone());
    Check {
        name: name.into(),
        status: Status::Ok,
        detail,
    }
}

fn extends_baseline(config_path: &str) -> bool {
    std::fs::read_to_string(config_path)
        .map(|s| s.contains("lazar-harness/linters") || s.contains("lint-baselines"))
        .unwrap_or(false)
}

// parseSelects is a best-effort reader for the ruff `select` list. It strips
// TOML line comments first, because real configs annotate every family, then
// reads the plain quoted-name subset.
fn parse_selects(config_path: &str) -> Vec<String> {
    let Ok(content) = std::fs::read_to_string(config_path) else {
        return vec![];
    };
    let no_comments: String = content
        .lines()
        .map(|l| match l.find('#') {
            Some(idx) => &l[..idx],
            None => l,
        })
        .collect::<Vec<_>>()
        .join(" ");
    let Some(start) = no_comments.find("select") else {
        return vec![];
    };
    let rest = &no_comments[start..];
    let Some(open) = rest.find('[') else {
        return vec![];
    };
    let Some(close) = rest[open..].find(']') else {
        return vec![];
    };
    rest[open + 1..open + close]
        .split(',')
        .map(|p| p.trim().trim_matches(|c| c == '"' || c == '\'').to_string())
        .filter(|s| !s.is_empty())
        .collect()
}

fn missing_families(repo_config: &str, opts: &resolve::Options) -> Vec<String> {
    let repo = parse_selects(repo_config);
    if repo.is_empty() {
        return vec![];
    }
    let baseline = parse_selects(&resolve::baseline_path(opts, "ruff.toml"));
    baseline.into_iter().filter(|f| !repo.contains(f)).collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use std::path::Path;

    const BASELINE: &str =
        "select = [\"E\", \"F\", \"I\", \"UP\", \"B\", \"SIM\", \"RET\", \"S\", \"RUF\"]\n";

    fn write(path: &str, content: &str) {
        fs::create_dir_all(Path::new(path).parent().unwrap()).unwrap();
        fs::write(path, content).unwrap();
    }

    fn unique(tag: &str) -> String {
        let thread = std::thread::current()
            .name()
            .unwrap_or("t")
            .replace("::", "-");
        std::env::temp_dir()
            .join(format!("hc-{}-{}-{}", tag, std::process::id(), thread))
            .to_string_lossy()
            .into_owned()
    }

    fn baseline_dir() -> String {
        let dir = std::path::PathBuf::from(unique("doctor-baseline"));
        write(&format!("{}/ruff.toml", dir.display()), BASELINE);
        write(&format!("{}/oxlint.json", dir.display()), "{}");
        dir.to_string_lossy().into()
    }

    fn has(report: &Report, name: &str, status: Status, part: &str) -> bool {
        report
            .checks
            .iter()
            .any(|c| c.name == name && c.status == status && c.detail.contains(part))
    }

    #[test]
    fn no_config_warns_with_baseline_message() {
        let repo = std::env::temp_dir().join("hc-doctor-noconfig");
        fs::create_dir_all(&repo).unwrap();
        let report = run(
            repo.to_str().unwrap(),
            &resolve::Options {
                baseline_dir: Some(baseline_dir()),
                ..Default::default()
            },
        );
        assert!(
            has(&report, "ruff-config", Status::Warn, "baseline applies"),
            "{report:?}"
        );
    }

    #[test]
    fn extending_repo_passes() {
        let repo = std::env::temp_dir().join("hc-doctor-extends");
        write(
            &format!("{}/ruff.toml", repo.display()),
            &format!("extend = \"/home/x/.config/lazar-harness/linters/ruff.toml\"\n{BASELINE}"),
        );
        let report = run(
            repo.to_str().unwrap(),
            &resolve::Options {
                baseline_dir: Some(baseline_dir()),
                ..Default::default()
            },
        );
        assert!(
            has(&report, "ruff-config", Status::Ok, "extends"),
            "{report:?}"
        );
        assert!(
            has(&report, "ruff-families", Status::Ok, "covers all"),
            "{report:?}"
        );
    }

    #[test]
    fn missing_families_reported() {
        let repo = std::env::temp_dir().join("hc-doctor-missing");
        write(
            &format!("{}/ruff.toml", repo.display()),
            &format!(
                "extend = \"~/.config/lazar-harness/linters/ruff.toml\"\nselect = [\"E\", \"F\"]\n"
            ),
        );
        let report = run(
            repo.to_str().unwrap(),
            &resolve::Options {
                baseline_dir: Some(baseline_dir()),
                ..Default::default()
            },
        );
        assert!(
            has(
                &report,
                "ruff-families",
                Status::Warn,
                "UP, B, SIM, RET, S, RUF"
            ),
            "{report:?}"
        );
    }

    #[test]
    fn own_config_not_extending_warns() {
        let repo = std::env::temp_dir().join("hc-doctor-own");
        write(&format!("{}/ruff.toml", repo.display()), BASELINE);
        let report = run(
            repo.to_str().unwrap(),
            &resolve::Options {
                baseline_dir: Some(baseline_dir()),
                ..Default::default()
            },
        );
        assert!(
            has(&report, "ruff-config", Status::Warn, "does not extend"),
            "{report:?}"
        );
    }
}
